File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int[] ans = new int[n-k+1];
        int idx = 0;

        Deque<Integer> dq = new LinkedList<>();
        for(int i = 0; i < n; ++i) {
            if(!dq.isEmpty() && dq.peekFirst() == i-k) dq.pollFirst(); // remove out-of-window element
            
            while(!dq.isEmpty() && nums[dq.peekLast()] <= nums[i]) dq.pollLast();
            dq.offerLast(i);
            if(i >= k-1) ans[idx++] = nums[dq.peekFirst()];
        }

        return ans;
    }
}

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        int i = 0, j = 0;
        deque<int> dq;
        vector<int> ans;
        while(j < n) {
            if(!dq.empty() && dq.front() == j-k) dq.pop_front();
            while(!dq.empty() && nums[dq.back()] <= nums[j]) dq.pop_back();
            dq.push_back(j);
            if(j >= k-1) ans.push_back(nums[dq.front()]);
            j++;
        }
        return ans;
    }
};